feat: Versioned timeouts CIP draft (#216)
celestiaorg/CIPs: cips/cip-26.md

| cip | 26 |
| - | - |
| title | Reduce block time to six seconds |

> | description | Proposal to reduce block time on celestia-app to 6 seconds, from 12. |

Reviewer: I'd rename the CIP to "versioned timeout", while the end goal is to reduce block time, but the actual change is that timeouts are now application version dependant, which would have a long-term effect from v3 on, and that is the timeouts will always be controlled by the application version (unless we change  the implementation in the future), and the timeouts in the configs won't be effective anymore.

Reviewer: Sure, done!

> | author | Josh Stein ([@jcstein](https://github.com/jcstein)), Rootul Patel ([@rootulp](https://github.com/rootulp))|

Reviewer: ```suggestion
| author | Josh Stein ([@jcstein](https://github.com/jcstein)), Rootul Patel ([@rootulp](https://github.com/rootulp))| Sanaz Taheri ([@staheri14](https://github.com/staheri14) )
```

| discussions-to | <https://forum.celestia.org/t/cip-decrease-block-time-to-6-seconds/1836> |
| status | Draft |
| type | Standards Track |
| category | Core |
| created | 2024-10-09 |

## Abstract

> This CIP proposes to reduce the block time on celestia-app to 6 seconds, from 12. This will double the throughput and reduce the time it takes for transactions to be finalized in half. This CIP also proposes to increase the `ttl-num-blocks` parameter in the mempool configuration from 5 to 12 to maintain consistency with the new block time.

Reviewer: >This will double the throughput and reduce the time it takes for transactions to be finalized in half

This is not entirely necessarily correct, in the sense that the users wont necessarily see 2x bytes per second after this change, for example, blocks can be empty or not full, hence the throughput would not be doubled. Also the same for finalization of txs, it depends on other factors, and while reducing block time has an impact, it does not exactly cut the finalization in half.

Author: Assuming blocks are being filled, this would double throughput (2MB/12s -> 2MB/6s) and cut down finalization from 12s to 6s? right? what would you suggest this changes to?

Reviewer: >Assuming blocks are being filled

That is exactly my point, throughput is a function of block size as well, not just block time, as you also mentioned in your comment "Assuming blocks are being filled".

Reviewer: > cut down finalization from 12s to 6s?

In addition to the block time, txs priority also contribute to their finality. So, it is a multifold problem, hence cutting block time does not necessarily make all txs to be finalized faster, in my pov.

> This CIP proposes to reduce the block time on celestia-app to 6 seconds, from 12. This will double the throughput and reduce the time it takes for transactions to be finalized in half. This CIP also proposes to increase the `ttl-num-blocks` parameter in the mempool configuration from 5 to 12 to maintain consistency with the new block time.

Reviewer: ```suggestion
This CIP proposes making timeouts application-version dependent. Starting from v3, timeouts will be controlled by the application version, rendering manual timeout configurations ineffective unless the implementation changes in the future.

This change enables automated block time adjustments, eliminating the need for validators to modify configurations manually, as the adjustments (if any) will occur automatically with each celestia-app version upgrade.

Updating the timeouts will naturally impact block time, block rate, and network throughput. For v3 of celestia-app, the timeout values are set to reduce the block time from 12 seconds to 6 seconds. This means block time is cut in half which, consequently, will nearly double the block rate and throughput (considering other network factors).

Additionally, this CIP proposes increasing the `ttl-num-blocks` parameter in the mempool configuration from 5 to 12 to align with the reduced block time.
```

## Motivation

The motivation for this CIP stems from a discussion in Core Devs Call 17, where it was proposed to reduce the block time to 6 seconds from 12 seconds.

## Specification

1. The block time in celestia-app SHOULD be reduced from 12 seconds to 6 seconds. Concretely, this implies decreasing `TimeoutCommit` to 4.2 seconds and `TimeoutPropose` to 3.5 seconds.

>     1. The `TimeoutCommit` and `TimeoutPropose` parameters were moved from local config parameters into versioned parameters controlled by the state machine to ensure consistency and correctness across different protocol versions.

Author: _from [comment](https://github.com/celestiaorg/CIPs/pull/216#pullrequestreview-2373407297)_
>  I think there was some complexity involved in moving the TimeoutCommit and TimeoutPropose from local config to state-machine managed so perhaps you can elaborate on the implementation details in this CIP?

would you want to change anything here @staheri14 ?

Reviewer: >I think there was some complexity involved in moving the TimeoutCommit and TimeoutPropose from local config to state-machine managed so perhaps you can elaborate on the implementation details in this CIP?

@rootulp Yes, there have been complications, we needed to change ABCI messages to include timeout info, do you mean elaborating on that aspect? or something else?

Author: I think he means elaborating it, but I also added line 29 after his feedback. does it look okay to you @staheri14 ?

>     1. The `TimeoutCommit` and `TimeoutPropose` parameters were moved from local config parameters into versioned parameters controlled by the state machine to ensure consistency and correctness across different protocol versions. This change allows the system to adapt dynamically to version-specific requirements, improving overall stability and interoperability during upgrades.

Reviewer 2: [nit] seems too AI generated

```suggestion
    1. The `TimeoutCommit` and `TimeoutPropose` parameters were moved from local config parameters into versioned parameters controlled by the state machine to ensure consistency and correctness across different protocol versions.
```

>     1. The change in block time MUST be implemented at a specific block height, which SHALL be determined and agreed upon by the Celestia community via on-chain signaling by validators.

Author: ```suggestion
```

>     1. The change in block time MUST be implemented at a specific block height, which SHALL be determined and agreed upon by the Celestia community via on-chain signaling by validators.

Reviewer: This sentence reads as if there is a separate signalling mechanism for block time change, however, afaik, signalling is a way to agree on a specific version of celestia-app to be used by validators, and since timeouts rely on the celestia-app version, when version transition happens so does the timeout update (if any). I'd suggest revising this sentence to make this point clear.

Reviewer: It would be also beneficial to add a link to the CIP containing the signalling mechanism.

Author: signaling mechanism is linked in meta CIP that this is included in, so wondering if we should take out this sentence entirely. wdyt @rootulp ?

Author: alternatively if you would like to propose a version of this sentence that makes more sense, would be happy to accept it too

    1. Celestia consensus nodes SHOULD update their software to accommodate this change prior to the agreed-upon block height.
    1. Client applications interacting with the Celestia network SHOULD be updated to account for the faster block time, particularly in areas related to transaction confirmation and block finality.

1. The default `ttl-num-blocks` parameter in the mempool configuration SHALL be increased from 5 to 12. This change is necessary to maintain consistency with the new block time and ensure that transactions remain in the mempool for a similar duration as before.
    1. Current default: `ttl-num-blocks = 5`
    1. New default: `ttl-num-blocks = 12`

>     1. This change SHALL NOT be implemented alongside the block time reduction. The default increase from 5 to 12 will occur when users upgrade to v3.0.0 and regenerate their config files. The block time reduction will happen one week later when the v2 to v3 activation height occurs. This approach ensures consistent behavior of the mempool across the network upgrade.

Author: ```suggestion
    1. This change SHALL NOT be implemented alongside the block time reduction. The default increase from 5 to 12 will occur when users upgrade to celestia-app v3.0.0 and regenerate their config files. The block time reduction will happen one week later when the v2 to v3 activation height occurs. This approach ensures consistent behavior of the mempool across the network upgrade.
```

>     1. This change SHALL NOT be implemented alongside the block time reduction. The default increase from 5 to 12 will occur when users upgrade to v3.0.0 and regenerate their config files. The block time reduction will happen one week later when the v2 to v3 activation height occurs. This approach ensures consistent behavior of the mempool across the network upgrade.

Reviewer: >This change SHALL NOT be implemented alongside the block time reduction. 

It is not immediately clear what is the order of operations here, and specifically what _upgrade_ means here:

>The default increase from 5 to 12 will occur when users upgrade to v3.0.0 and regenerate their config files.

Nodes switch their binaries to v3? or when signalling mechanism is conducted and the results indicates an app version upgrade to v3 is going to take place?

Author: This was added after this feedback https://github.com/celestiaorg/CIPs/pull/216#discussion_r1803240986

>     1. All validator nodes MUST update their configuration files to reflect this new `ttl-num-blocks` value before the agreed-upon implementation block height.

Reviewer 2: we currently don't enforce it

```suggestion
    1. All validator nodes SHOULD update their configuration files to reflect this new `ttl-num-blocks` value before the agreed-upon implementation block height.
```

1. Documentation and APIs related to block time and block production MUST be updated to reflect these changes.

## Rationale

The rationale for this change is to increase the throughput of the Celestia network by doubling the number of blocks produced per unit of time. This will reduce the time it takes for transactions to be finalized and improve the overall user experience on the network.

The increase in `ttl-num-blocks` from 5 to 12 is necessary to maintain consistent mempool behavior with the new block time. This change ensures that transactions remain in the mempool for approximately 72 seconds (12 blocks times 6 seconds), which closely matches the previous behavior of about 60 seconds (5 blocks times 12 seconds).

## Backwards Compatibility

> This is a breaking network upgrade and will require all participants to update their software to accommodate the new block time and `ttl-num-blocks`. The change in block time will not be backward compatible with the existing network, and all participants MUST be compliant by the agreed-upon implementation block height.

Reviewer: I am not sure if this is a breaking upgrade? would like to cc @evan-forbes for his pov
When we are talking about backward compatibility, does it mean that a validator running v2 binary wouldn't be able to participate in a network running v3 binary? If this is the case, then I don't see how it is not backward compatible; validators running v2 will only have different timeouts, and yes, they may send out their proposal late since their timeout commit is longer than others, but it does not seem like a breaking upgrade.

Author: > When we are talking about backward compatibility, does it mean that a validator running v2 binary wouldn't be able to participate in a network running v3 binary?
My understanding is that this is breaking behavior? i.e. I can't run a v2 binary once v3 goes live?

Author: cc: @cmwaters here as well, I think now I understand maybe this isn't breaking, but not sure.

## Test Cases

This will be tested on Arabica devnet and Mocha testnet before going live on Celestia Mainnet Beta.

## Security Considerations

While the reduction in block time itself does not introduce significant new security risks to the network, there are important considerations:

1. Participants should ensure that their systems are capable of handling the increased throughput from faster block times.

> 1. The increase of `ttl-num-blocks` from 5 to 12 is crucial for maintaining the security and efficiency of the mempool:

Reviewer: [question] Shouldn’t we increase the mempool size? My intuition is that we’re likely processing twice as many transactions in the same amount of time as before, so it might be worth considering a larger mempool to accommodate the increased volume.

Author: I'm resolving this based on [slack discussion](https://celestia-team.slack.com/archives/C02PFFCGMNW/p1729115035368559), but please feel free to reopen

    1. It prevents premature removal of valid transactions, reducing the risk of unintended exclusion from blocks.
    1. Without this adjustment, transactions would be pruned from the mempool after only 30 seconds, potentially leading to increased transaction failures and a poor user experience.

> 1. Validators and node operators must update their configurations to reflect the new `ttl-num-blocks` value to maintain network consistency and security.

Reviewer 2: ```suggestion
1. Validators and node operators should update their configurations to reflect the new `ttl-num-blocks` value to maintain network consistency and security.
```

This change enables automated block time adjustments, eliminating the need for validators to modify configurations manually, as the adjustments (if any) will occur automatically with each celestia-app version upgrade.

Updating the timeouts will naturally impact block time, block rate, and network throughput. For v3 of celestia-app, the timeout values are set to reduce the block time from 12 seconds to 6 seconds. This means block time is cut in half which, consequently, will nearly double the block rate and throughput (considering other network factors).

Additionally, this CIP proposes increasing the `ttl-num-blocks` parameter in the mempool configuration from 5 to 12 to align with the reduced block time.

## Motivation

The motivation for this CIP stems from a discussion in Core Devs Call 17, where it was proposed to reduce the block time to 6 seconds from 12 seconds.

## Specification

1. The block time in celestia-app SHOULD be reduced from 12 seconds to 6 seconds. Concretely, this implies decreasing `TimeoutCommit` to 4.2 seconds and `TimeoutPropose` to 3.5 seconds.

>     1. The `TimeoutCommit` and `TimeoutPropose` parameters were moved from local config parameters into versioned parameters controlled by the state machine to ensure consistency and correctness across different protocol versions.

Reviewer: I'd suggest the following revision:

```suggestion
    1. The `TimeoutCommit` and `TimeoutPropose` parameters were moved from local config parameters into versioned parameters controlled by the state machine. The timeouts will be managed by the application and communicated with Celestia-core through the following ABCI interfaces: `InitChain`, `EndBlock`, and `Info`, now extended with `TimeoutsInfo`, which encapsulates `TimeoutPropose` and `TimeoutCommit`. The timeouts obtained through these interfaces are utilized by the Celestia-core side as needed.
```

    1. All validator nodes SHOULD update their configuration files to reflect this new `ttl-num-blocks` value before the agreed-upon implementation block height.

1. Documentation and APIs related to block time and block production MUST be updated to reflect these changes.

## Rationale

The rationale for this change is to increase the throughput of the Celestia network by doubling the number of blocks produced per unit of time. This will reduce the time it takes for transactions to be finalized and improve the overall user experience on the network.

The increase in `ttl-num-blocks` from 5 to 12 is necessary to maintain consistent mempool behavior with the new block time. This change ensures that transactions remain in the mempool for approximately 72 seconds (12 blocks times 6 seconds), which closely matches the previous behavior of about 60 seconds (5 blocks times 12 seconds).

## Backwards Compatibility

> This is a breaking network upgrade and will require all participants to update their software to accommodate the new block time and `ttl-num-blocks`. The change in block time will not be backward compatible with the existing network, and all participants MUST be compliant by the agreed-upon implementation block height.

Author: ```suggestion
This upgrade requires all participants to update their software to v3 to accommodate the new block time and `ttl-num-blocks`. Nodes running older versions may not function correctly with the new network parameters. All validators and node operators should update to v3 before the agreed-upon implementation block height to ensure network consistency and optimal performance.
```

| cip | 26 |
| - | - |

> | title | Reduce block time to six seconds |

Author: ```suggestion
| title | Versioned timeout |
```

> | description | Proposal to reduce block time on celestia-app to 6 seconds, from 12. |

Author: ```suggestion
| description | Timeouts are now controlled by the application version. |
```
